clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1202,6 +1262,87 @@ class SyclKernelFieldChecker : public SyclKernelFieldHandler {
   }
 };
 
+// A type to check the validity of accessing accessor/sampler/stream
+// types as kernel parameters inside union.
+class SyclKernelUnionChecker : public SyclKernelFieldHandler {
+  int UnionCount = 0;
+  bool IsInvalid = false;
+  DiagnosticsEngine &Diag;
+
+public:
+  SyclKernelUnionChecker(Sema &S)
+      : SyclKernelFieldHandler(S), Diag(S.getASTContext().getDiagnostics()) {}
+  bool isValid() { return !IsInvalid; }
+  static constexpr const bool VisitUnionBody = true;
+
+  bool enterUnion(const CXXRecordDecl *RD, FieldDecl *FD) {
+    ++UnionCount;
+    return true;
+  }
+
+  bool leaveUnion(const CXXRecordDecl *RD, FieldDecl *FD) {
+    --UnionCount;
+    return true;
+  }
+
+  bool handleSyclAccessorType(FieldDecl *FD, QualType FieldTy) final {
+    if (UnionCount) {
```

Comment:
This is a NIT, so don't redo the patch just for this (but please do if you have to push another patch), but I'd prefer the body of this to be extracted into a private function.  Something like: 
`bool checkType(SourceLocation Loc, QualType Ty) {
   if (UnionCount) {
      IsInvalid = true;
      DiagReport(Loc, diag::err_bad_union_kernel_param_members) << Ty);
   }
   return isValid();
}`

Then, all of these other functions simply become:

`return checkType(FD->getLocation() /* or BS.getBeginLoc()*/, FieldTy);`

I want this, since I believe it will make the 'printing the diagnostic trail' much easier.

Reply:
Done.